crates/wasmparser/src/binary_reader.rs

```diff
@@ -1329,19 +1336,74 @@ impl<'a> BinaryReader<'a> {
         Ok(self.buffer[self.position])
     }
 
-    fn val_type_from_byte(byte: u8) -> Option<ValType> {
-        match byte {
-            0x7F => Some(ValType::I32),
-            0x7E => Some(ValType::I64),
-            0x7D => Some(ValType::F32),
-            0x7C => Some(ValType::F64),
-            0x7B => Some(ValType::V128),
-            0x70 => Some(ValType::FuncRef),
-            0x6F => Some(ValType::ExternRef),
-            _ => None,
+    fn read_heap_type(&mut self) -> Result<HeapType> {
+        match self.peek()? {
+            0x70 => {
+                self.position += 1;
+                Ok(HeapType::Func)
+            }
+            0x6F => {
+                self.position += 1;
+                Ok(HeapType::Extern)
+            }
+            x if 0x80 & x == 0 => {
+                let idx = self.read_var_s33()?;
+                if idx < 0 || idx > (std::u16::MAX as i64) {
+                    return Err(BinaryReaderError::new(
+                        "invalid function heap type",
+                        self.original_position(),
+                    ));
+                }
+                Ok((idx as u32).try_into().unwrap())
+            }
+            x => Err(BinaryReaderError::new(
+                format!("unknown heap type subopcode: 0x{:x}", x),
+                self.original_position(),
+            )),
         }
     }
 
+    fn maybe_read_val_type(&mut self) -> Result<Option<ValType>> {
+        Ok(match self.peek()? {
+            0x7F => {
+                self.position += 1;
+                Some(ValType::I32)
+            }
+            0x7E => {
+                self.position += 1;
+                Some(ValType::I64)
+            }
+            0x7D => {
+                self.position += 1;
+                Some(ValType::F32)
+            }
+            0x7C => {
+                self.position += 1;
+                Some(ValType::F64)
+            }
+            0x7B => {
+                self.position += 1;
+                Some(ValType::V128)
+            }
+            0x70 => {
+                self.position += 1;
+                Some(ValType::Ref(FUNC_REF))
+            }
+            0x6F => {
+                self.position += 1;
+                Some(ValType::Ref(EXTERN_REF))
+            }
+            x @ (0x6B | 0x6C) => {
+                self.position += 1;
+                Some(ValType::Ref(RefType {
+                    nullable: x == 0x6C,
+                    heap_type: self.read_heap_type()?,
+                }))
+            }
+            _ => None,
```

Comment:
I think that this, and transitively `read_heap_type` aren't quite right because they need to have the fallback case of any `s33`-encoded value. To test out the exponential behavior I pointed out below I created a module with 100 types but the module failed to parse I think due to what I think is an issue in `read_heap_type` (although I'm not 100% sure).

I think it might be good to add some test cases with a type index of 256 and place that into various places (e.g. block types, params, results, `(ref $foo)`, etc) and make sure that the binary encoding/decoding works everywhere.